### PR TITLE
feature: supported specifying environment variables when spawning the…

### PR DESCRIPTION
**Pull request by spacewander**
… sub-process.

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-resty-core project.

### CODE REVIEW

**Comment by spacewander** [comment]
This is the sister PR of https://github.com/openresty/lua-nginx-module/pull/1513

File: t/pipe.t [hunk 1]
```diff
@@ -1546,3 +1546,57 @@ exit
...
+            spawn(environ)
+            table.clear(environ)
+            spawn(environ)
+            spawn("TEST_ENV=1")
+        }
+    }
+--- response_body
+bad environ option in index 2: string expected, got number
+bad environ option: non-empty table expected
+bad environ option: table expected, got string
```
Comment (thibaultcha) [inline t/pipe.t]: Should we add a test with a `nil` hole somewhere in the `environ` option table?

File: lib/ngx/pipe.lua [hunk 2]
```diff
@@ -566,6 +567,38 @@ do
                     error("bad buffer_size option", 2)
                 end
             end
+
+            if opts.environ then
+                environ = opts.environ
+                local environ_type = type(environ)
+                if environ_type ~= 'table' then
```
Comment (thibaultcha) [inline lib/ngx/pipe.lua]: style: could we stick to double quotes to stay consistent with the rest of this function's body style?

File: lib/ngx/pipe.lua [hunk 3]
```diff
@@ -566,6 +567,38 @@ do
...
+                    error("bad environ option: table expected, got "
+                          .. environ_type, 2)
+                end
+
+                local nenv = 0
+
+                for i, env in ipairs(environ)  do
```
Comment (thibaultcha) [inline lib/ngx/pipe.lua]: style: extra space before `do`

File: lib/ngx/pipe.lua [hunk 4]
```diff
@@ -566,6 +567,38 @@ do
...
+                    if env == nil then
+                        break
+                    end
+
+                    nenv = nenv + 1
+
+                    local env_type = type(env)
+                    if env_type ~= 'string' then
+                        error("bad environ option in index " .. i
+                              .. ": string expected, got " .. env_type, 2)
```
[the remark this answers is not in the document]
Comment (spacewander) [reply lib/ngx/pipe.lua]: @thibaultcha 
Environment variable is passed like this format: `KEY=VALUE`. So a `number` argument should not be a valid one.
Maybe we could be more strict, check if there is a `=` in the argument.

[the remark this answers is not in the document]
Comment (thibaultcha) [reply lib/ngx/pipe.lua]: My bad, while reviewing this particular line I had a hash construct in mind for the `environ` option, which is wrong of course, since the current implementation is an array. Long day.

Such validation is a good idea imho, and adding some test cases with malformed environment strings as well.

**Comment by spacewander** [comment]
@thibaultcha 
Updated and rebased.

**Comment by thibaultcha** [comment]
@spacewander Thanks. I made a few more adjustments locally and added additional tests, but waiting for @agentzh to answer a process question I had before merging it.

**Comment by thibaultcha** [comment]
Merged!
